docs/cheat-sheets/python-command-injection.mdx

```diff
@@ -0,0 +1,649 @@
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.dangerous-system-call" />
+
+### 1.D. Using **os** module to spawn a process
+
+The `os` module allows executing the program path in a new process. Variations of spawn method: `spawnl`, `spawnle`, `spawnlp`, `spawnlpe`, `spawnv`, `spawnve`, `spawnvp`, `spawnvpe`, `posix_spawn` and `posix_spawnp` are intended for spawning a process with a program passed as a string argument. Allowing spawning arbitrary programs or running shell processes with arbitrary arguments may end up in a command injection vulnerability.
+
+Example:
+
+```python
+import os
+
+# Vulnerable
+user_input = "/foo/bar" # value supplied by user
+os.spawnlpe(os.P_WAIT, user_input, ["-a"], os.environ)
+
+# Vulnerable
+user_input = "cat /etc/passwd" # value supplied by user
+os.spawnve(os.P_WAIT, "/bin/bash", ["-c", user_input], os.environ)
+```
+
+#### References:
+
+- [**os** module documentation](https://docs.python.org/3/library/os.html)
+
+#### Mitigation
+
+Do not let user input in `spawn` methods. Alternatively,
+
+- Always try to use internal Python API (if it exists) instead of running an OS command.
+- Don’t pass user-controlled input.
+- Use `shlex.split` to correctly parse command string into an array and `shlex.quote` to correctly sanitize input as a command-line parameter.
+- Avoid running `sh` as a command with arguments, if it is not possible - strip everything except alphanumeric characters from an input provided for the command string and arguments.
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.dangerous-spawn-process.dangerous-spawn-process" />
+
+### 1.E. Replacing currnet process with **exec**
+
+Execution methods of the `os` module are intended to execute a new program, replacing the current process. All available methods are `execl`, `execle`, `execlp`, `execlpe`, `execv`, `execve`, `execvp`, `execvpe`. Allowing running arbitrary programs or running shell processes with arbitrary arguments may end up in a command injection vulnerability.
+
+Example:
+
+```python
+import os
+
+# Vulnerable
+user_input = "/evil/code" # value supplied by user
+os.execl(user_input, '/foo/bar', '--do-smth')
+
+# Vulnerable
+user_input = "cat /etc/passwd" # value supplied by user
+os.execve("/bin/bash", ["/bin/bash", "-c", user_input], os.environ)
+```
+
+#### References:
+
+- [**os** module documentation](https://docs.python.org/3/library/os.html)
+
+#### Mitigation
+
+Do not let user input in `exec` methods. Alternatively,
+
+- Always try to use internal Python API (if it exists) instead of running an OS command.
+- Don’t pass user-controlled input.
+- Use `shlex.split` to correctly parse command string into an array and `shlex.quote` to correctly sanitize input as a command-line parameter.
+- Avoid running `sh` as a command with arguments, if it is not possible - strip everything except alphanumeric characters from an input provided for the command string and arguments.
```

Comment:
```suggestion
- Avoid running `sh` as a command with arguments. If it’s not possible to avoid, strip everything except alphanumeric characters from an input provided for the command string and arguments.
```